src/shims/unix/linux/eventfd.rs

```diff
@@ -217,8 +179,205 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+
+    waiter.sort();
+    waiter.dedup();
+    for thread_id in waiter {
+        ecx.unblock_thread(thread_id, BlockReason::Eventfd)?;
+    }
+
+    // Tell userspace how many bytes we read.
+    ecx.write_int(buf_place.layout.size.bytes(), dest)
+}
+
+/// Add the user-supplied value to the current counter.
+fn eventfd_write<'tcx>(
+    num: u64,
+    buf_place: &MPlaceTy<'tcx>,
+    dest: &MPlaceTy<'tcx>,
+    weak_eventfd: WeakFileDescriptionRef,
+    ecx: &mut MiriInterpCx<'tcx>,
+) -> InterpResult<'tcx> {
+    let Some(eventfd_ref) = weak_eventfd.upgrade() else {
+        throw_unsup_format!("eventfd FD got closed while blocking.")
+    };
+
+    // Since we pass the weak file description ref to the function, it is guaranteed to be
+    // an eventfd file description.
+    let eventfd = eventfd_ref.downcast::<Event>().unwrap();
+
+    // Future `read` calls will synchronize with this write, so update the FD clock.
+    ecx.release_clock(|clock| {
+        eventfd.clock.borrow_mut().join(clock);
+    });
+
+    // When this function is called, the addition is guaranteed to not exceed u64::MAX - 1.
+    let new_count = eventfd.counter.get().checked_add(num).unwrap();
+    eventfd.counter.set(new_count);
+
+    // When any of the event happened, we check and update the status of all supported event
+    // types for current file description.
+    ecx.check_and_update_readiness(&eventfd_ref)?;
+
+    // Unblock *all* threads previously blocked on `read`.
+    // We need to store the blocked thread ids and unblock them together to prevent BorrowMutError
+    // panic because when unblock_thread is called, blocked_read_tid will be mutably borrowed again.
+    let mut waiter = Vec::new();
+    let mut blocked_read_tid = eventfd.blocked_read_tid.borrow_mut();
+    while let Some(tid) = blocked_read_tid.pop() {
+        waiter.push(tid);
+    }
+    drop(blocked_read_tid);
+
+    waiter.sort();
+    waiter.dedup();
+    for thread_id in waiter {
+        ecx.unblock_thread(thread_id, BlockReason::Eventfd)?;
+    }
+
+    // Return how many bytes we wrote.
+    ecx.write_int(buf_place.layout.size.bytes(), dest)
+}
+
+/// Block thread if the value addition will exceed u64::MAX -1,
+/// else just add the user-supplied value to current counter.
+fn check_write_value_and_block_thread<'tcx>(
+    num: u64,
+    buf_place: MPlaceTy<'tcx>,
+    dest: &MPlaceTy<'tcx>,
+    weak_eventfd: WeakFileDescriptionRef,
+    ecx: &mut MiriInterpCx<'tcx>,
+) -> InterpResult<'tcx> {
+    let Some(eventfd_ref) = weak_eventfd.upgrade() else {
+        throw_unsup_format!("eventfd FD got closed while blocking.")
+    };
+
+    // Since we pass the weak file description ref, it is guaranteed to be
+    // an eventfd file description.
+    let eventfd = eventfd_ref.downcast::<Event>().unwrap();
+
+    match eventfd.counter.get().checked_add(num) {
+        Some(_new_count @ 0..=MAX_COUNTER) => {
+            return eventfd_write(num, &buf_place, dest, weak_eventfd, ecx);
```

Comment:
Why is this a separate function? And why the name? `check_write_value_and_block_thread` is actually implementing the full up, `eventfd_write` just a part of it.

I'd suggest to inline `eventfd_write` here, and rename `check_write_value_and_block_thread` to `eventfd_write`.